File: app/src/main/java/life/wanderinglocal/WLTimeLineAdapter.java
package life.wanderinglocal;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import static com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions.withCrossFade;

public class WLTimeLineAdapter extends RecyclerView.Adapter<WLTimeLineAdapter.YelpBusiness> {
    private int displayHeight = -1;

    public int getDisplayHeight() {
        if (displayHeight == -1) {
            DisplayMetrics displayMetrics = new DisplayMetrics();
            WindowManager windowManager =
                    (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
            windowManager.getDefaultDisplay().getMetrics(displayMetrics);
            displayHeight = displayMetrics.heightPixels;
        }
        return displayHeight;
    }

    private List<WLTimelineEntry> data = new ArrayList<>();
    private Context context;

    public WLTimeLineAdapter(Context context) {
        this.context = context;
    }

    public void addData(WLTimelineEntry yd) {
        data.add(yd);
        notifyDataSetChanged();
    }

    public void setData(List<WLTimelineEntry> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public YelpBusiness onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.yelp_business_row, parent, false);
        YelpBusiness yb = new YelpBusiness(v);
        return yb;
    }

    @Override
    public void onBindViewHolder(@NonNull YelpBusiness holder, int position) {
        final WLTimelineEntry timelineEntry = data.get(position);
        holder.businessNameTextView.setText(timelineEntry.getBusinessName());
        holder.businessAddressTextView.setText(timelineEntry.getLocationString());
        holder.businessAddressTextView.setOnClickListener(view -> {
            Uri locationUri = Uri.parse(String.format("http://maps.google.com/maps?q=" + timelineEntry.getLocationString()));
            Intent intent = new Intent(Intent.ACTION_VIEW, locationUri);
            context.startActivity(intent);
        });
        holder.imgView.setImageBitmap(timelineEntry.getBmp());

        ViewGroup.LayoutParams params = holder.imgView.getLayoutParams();
        params.height = getDisplayHeight() / 3;
        params.width = getDisplayHeight() / 3;
        holder.imgView.setLayoutParams(params);
        Glide.with(context).load(timelineEntry.imageUrl)
                .transition(withCrossFade())
                .centerCrop()
                        .skipMemoryCache(true)
                        .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .into(holder.imgView);

        holder.ratingBar.setNumStars(5);
        holder.ratingBar.setRating((float) timelineEntry.getRating());
        holder.itemView.setOnClickListener(view -> {
            if (timelineEntry.yelpUrl == null) return;
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.parse(timelineEntry.yelpUrl));
            context.startActivity(intent);
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    class YelpBusiness extends RecyclerView.ViewHolder {
        RatingBar ratingBar;
        ImageView imgView;
        TextView businessNameTextView;
        TextView businessAddressTextView;

        public YelpBusiness(@NonNull View itemView) {
            super(itemView);
            this.imgView = itemView.findViewById(R.id.businessImg);
            this.businessNameTextView = itemView.findViewById(R.id.businessName);
            this.ratingBar = itemView.findViewById(R.id.businessRatingBar);
            this.businessAddressTextView = itemView.findViewById(R.id.businessAddress);
        }
    }
}

File: app/src/main/java/life/wanderinglocal/WLTimeLineAdapter.kt
package life.wanderinglocal

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.DisplayMetrics
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import io.reactivex.rxjava3.subjects.PublishSubject
import io.reactivex.rxjava3.subjects.Subject
import life.wanderinglocal.WLTimeLineAdapter.YelpBusiness
import java.util.*

class WLTimeLineAdapter(private val context: Context) : RecyclerView.Adapter<YelpBusiness>() {
    val itemClickedSubject: Subject<WLTimelineEntry> = PublishSubject.create()
    var displayHeight = -1
        get() {
            if (field == -1) {
                val displayMetrics = DisplayMetrics()
                val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
                windowManager.defaultDisplay.getMetrics(displayMetrics)
                field = displayMetrics.heightPixels
            }
            return field
        }
    private var data: List<WLTimelineEntry> = ArrayList()

    /**
     * Replace with Google Paging Library
     */
    fun addData(yd: WLTimelineEntry) {
    }

    fun setData(data: List<WLTimelineEntry>) {
        this.data = data
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): YelpBusiness {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.yelp_business_row, parent, false)
        return YelpBusiness(v)
    }

    override fun onBindViewHolder(holder: YelpBusiness, position: Int) {
        val timelineEntry = data[position]
        holder.businessNameTextView.text = timelineEntry.businessName
        holder.businessAddressTextView.text = timelineEntry.locationString
        holder.businessAddressTextView.setOnClickListener { view: View? ->
            val locationUri = Uri.parse(String.format("http://maps.google.com/maps?q=" + timelineEntry.locationString))
            val intent = Intent(Intent.ACTION_VIEW, locationUri)
            context.startActivity(intent)
        }
        holder.imgView.setImageBitmap(timelineEntry.bmp)
        val params = holder.imgView.layoutParams
        params.height = displayHeight / 3
        params.width = displayHeight / 3
        holder.imgView.layoutParams = params
        Glide.with(context).load(timelineEntry.imageUrl)
                .transition(DrawableTransitionOptions.withCrossFade())
                .centerCrop()
                .skipMemoryCache(true)
                .diskCacheStrategy(DiskCacheStrategy.AUTOMATIC)
                .into(holder.imgView)
        holder.ratingBar.numStars = 5
        holder.ratingBar.rating = timelineEntry.rating.toFloat()
        holder.itemView.setOnClickListener { view: View? ->
            itemClickedSubject.onNext(timelineEntry)
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    class YelpBusiness(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var ratingBar: RatingBar
        var imgView: ImageView
        var businessNameTextView: TextView
        var businessAddressTextView: TextView

        init {
            imgView = itemView.findViewById(R.id.businessImg)
            businessNameTextView = itemView.findViewById(R.id.businessName)
            ratingBar = itemView.findViewById(R.id.businessRatingBar)
            businessAddressTextView = itemView.findViewById(R.id.businessAddress)
        }
    }

}